// @google-cloud/bigquery-storage/build/src/index.js
"use strict";
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.adapt = exports.protos = exports.WriterClient = exports.managedwriter = exports.BigQueryWriteClient = exports.BigQueryStorageClient = exports.v1beta1 = exports.BigQueryReadClient = exports.v1 = void 0;
const v1 = require("./v1");
exports.v1 = v1;
const v1beta1 = require("./v1beta1");
exports.v1beta1 = v1beta1;
const managedwriter = require("./managedwriter");
exports.managedwriter = managedwriter;
const BigQueryReadClient = v1.BigQueryReadClient;
exports.BigQueryReadClient = BigQueryReadClient;
const BigQueryWriteClient = v1.BigQueryWriteClient;
exports.BigQueryWriteClient = BigQueryWriteClient;
const BigQueryStorageClient = v1beta1.BigQueryStorageClient;
exports.BigQueryStorageClient = BigQueryStorageClient;
const WriterClient = managedwriter.WriterClient;
exports.WriterClient = WriterClient;
// For compatibility with JavaScript libraries we need to provide this default export:
// tslint:disable-next-line no-default-export
exports.default = {
    v1,
    BigQueryReadClient,
    BigQueryWriteClient,
    managedwriter,
    WriterClient,
};
const protos = require("../protos/protos");
exports.protos = protos;
const adapt = require("./adapt");
exports.adapt = adapt;
// Add extra protobufjs definitions.
// When importing protobufjs/ext/descriptor package, it monkey patches some methods
// that we use in this package. We need to manually declare some of those
// methods that we use to make the Typescript compiler happy.
// There are some open issues around that. After they are fixed, we can remove this:
// * https://github.com/protobufjs/protobuf.js/issues/1499
// * https://github.com/protobufjs/protobuf.js/issues/1149
require("./protobuf");
